cip/CIP2016-06-14-Redefine-comparability-and-equality-as-well-as-orderability-and-equivalence.adoc

```diff
@@ -0,0 +1,367 @@
+
+- General rules
+  * Values are only comparable within their most specific type (except for numbers, see below).
+  * Equal values are grouped together.
+- Numbers
+  * Integers are compared numerically in ascending order.
+  * Floats (excluding `NaN` and the Infinities) are compared numerically in ascending order.
+  * Numbers of different types (excluding `NaN` and the Infinities) are compared to each other as if both numbers would have been coerced to arbitrary precision big decimals (currently outside the Cypher type system) before comparing them with each other numerically in ascending order.
+  * Positive infinity is of type `FLOAT`, equal to itself and larger than any other number (excluding `NaN`).
+  * Negative infinity is of type `FLOAT`, equal to itself and smaller than any other number (excluding `NaN`).
+  * `NaN` is <<incomparable>>.
+  * Numbers are <<incomparable>> to any value that is not also a number.
+- Booleans
+  * Booleans are compared such that `false` is less than `true`.
+  * Booleans are <<incomparable>> to any value that is not also a boolean.
+- Strings
+  * Strings are compared in dictionary order, i.e. characters are compared pairwise in ascending order from the start of the string to the end.
+  Characters missing in a shorter string are considered to be smaller than any other character.
+  For example, `'a' < 'aa'`.
+  * Strings are <<incomparable>> to any value that is not also a string.
+- Lists
+  * Lists are compared in dictionary order, i.e. list elements are compared pairwise in ascending order from the start of the list to the end. Elements missing in a shorter list are considered to be smaller than any other value.
+  For example, `[1] < [1, 0]`.
+  * Lists containing `null` values are <<incomparable>>. For example, `[1] >= [1, null]` evaluates to `null`.
+  * Lists are <<incomparable>> to any value that is not also a list.
+- Maps
+  * [[regular-maps,regular maps]]Regular maps
+  ** The comparison order for maps is unspecified and left to implementations.
+  ** The comparison order for maps must align with the <<equality-def,equality semantics>> outlined below.
+  In consequence, any map that contains an entry that maps its key to a `null` value is <<incomparable>>.
+  For exampe, `{a: 1} \<= {a: 1, b: null}` evaluates to `null`.
+  ** Regular maps are <<incomparable>> to any value that is not also a regular map.
+  * Nodes
+  ** The comparison order for nodes is unspecified and left to implementations.
+  ** Nodes are <<incomparable>> to any value that is not also a node.
+  * Relationships
+  ** The comparison order for relationships is unspecified and left to implementations.
+  ** Relationships are <<incomparable>> to any value that is not also a relationship.
+- Paths
+  ** Paths are compared as if they were a list of alternating nodes and relationships of the path from the start node to the end node.
```

Comment:
The precise way in which nodes and relationships is left to the implementation however an implementation is required to define an order for them (I now made this more clear in the text). Paths simply piggyback on that.

Doing so, gives a nice lexicographic order on paths, i.e. you see them in some traversal like order when sorting them. I think that is useful.